### PR TITLE
Table: maps, zips & more builtins

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
1. Introduces zip operations into the API.
2. Refactors the way we define builtin ops, making it slightly less obnoxious
3. Introduces more builtin ops

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: table/src/main/java/org/enso/table/data/column/storage/BoolStorage.java [hunk 1]
```diff
@@ -133,4 +121,138 @@ public Storage countMask(int[] counts, int total) {
...
+                    out.or(storage.values);
+                    negated = false;
+                  }
+                  return new BoolStorage(out, missing, storage.size, negated);
+                } else {
+                  throw new UnexpectedColumnTypeException("Boolean");
+                }
+              }
+            });
+    return ops;
```
Comment (iamrecursion) [inline table/src/main/java/org/enso/table/data/column/storage/BoolStorage.java:256]: What do you think about adding `xor`? It's a boolean op I used fairly frequently back at Bloomberg for data stuff.

File: table/src/main/java/org/enso/table/data/column/storage/DoubleStorage.java [hunk 2]
```diff
@@ -134,4 +122,88 @@ public Storage countMask(int[] counts, int total) {
...
+            })
+        .add(
+            new DoubleNumericOp(Ops.MUL) {
+              @Override
+              protected double doDouble(double a, double b) {
+                return a * b;
+              }
+            })
+        .add(
+            new DoubleNumericOp(Ops.DIV) {
```
Comment (iamrecursion) [inline table/src/main/java/org/enso/table/data/column/storage/DoubleStorage.java:154]: Missing `%`.

File: table/src/main/java/org/enso/table/data/column/storage/LongStorage.java [hunk 3]
```diff
@@ -130,4 +122,133 @@ public Storage countMask(int[] counts, int total) {
...
+                return in * arg;
+              }
+
+              @Override
+              public long doLong(long in, long arg) {
+                return in * arg;
+              }
+            })
+        .add(
+            new LongNumericOp(Ops.DIV, true) {
```
Comment (iamrecursion) [inline table/src/main/java/org/enso/table/data/column/storage/LongStorage.java:181]: Missing `%`.

File: table/src/main/java/org/enso/table/error/UnexpectedTypeException.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+/** An error thrown when a type error is encountered. */
+public class UnexpectedTypeException extends RuntimeException {
+  private final String expected;
+
+  /**
+   * Creates a new instance of this error.
+   *
+   * @param expected the expected type description
+   */
+  public UnexpectedTypeException(String expected) {
```
Comment (iamrecursion) [inline table/src/main/java/org/enso/table/error/UnexpectedTypeException.java:12]: Is there a path at all to replacing this with the language's `Type_Error` atom?

**Review by iamrecursion — APPROVED** [review]
This is fantastic. Super well-designed, and very cleanly implemented.

File: table/src/main/java/org/enso/table/data/column/storage/BoolStorage.java (hunk 1 above)
Comment (kustosz) [reply table/src/main/java/org/enso/table/data/column/storage/BoolStorage.java:256]: But we have xor already! It's just that it's called `!=` 😛

File: table/src/main/java/org/enso/table/error/UnexpectedTypeException.java (hunk 4 above)
Comment (kustosz) [reply table/src/main/java/org/enso/table/error/UnexpectedTypeException.java:12]: Not really, if we want to go in that direction, we'll have to catch these in Enso and rethrow as Type_Error

**Pull request merged**